resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/DriverInitContainerComponentsProvider.scala

```diff
@@ -128,28 +129,36 @@ private[spark] class DriverInitContainerComponentsProviderImpl(
           maybeResourceStagingServerInternalTrustStorePassword,
           maybeResourceStagingServerInternalTrustStoreType,
           INIT_CONTAINER_SECRET_VOLUME_MOUNT_PATH)
-    }
-    new SparkInitContainerConfigMapBuilderImpl(
+      }
+      new SparkInitContainerConfigMapBuilderImpl(
         sparkJars,
         sparkFiles,
         jarsDownloadPath,
         filesDownloadPath,
         configMapName,
         configMapKey,
         submittedDependencyConfigPlugin)
+    } else new SparkInitContainerEmptyConfigMapBuilderImpl()
   }
 
   override def provideContainerLocalizedFilesResolver(): ContainerLocalizedFilesResolver = {
     new ContainerLocalizedFilesResolverImpl(
         sparkJars, sparkFiles, jarsDownloadPath, filesDownloadPath)
   }
+  val containerLocalizedFilesResolver = provideContainerLocalizedFilesResolver()
+  val resolvedSparkJars = containerLocalizedFilesResolver.resolveSubmittedSparkJars()
+  val resolvedSparkFiles = containerLocalizedFilesResolver.resolveSubmittedSparkFiles()
+  // Bypass init-containers if `spark.jars` and `spark.files` is empty or only has `local://` URIs
+  private val needInitContainer = (resolvedSparkJars ++ resolvedSparkFiles).exists { e =>
+    e.nonEmpty && !e.startsWith("local://")
+  }
 
   override def provideExecutorInitContainerConfiguration(): ExecutorInitContainerConfiguration = {
-    new ExecutorInitContainerConfigurationImpl(
+    if (needInitContainer) new ExecutorInitContainerConfigurationImpl(
```

Comment:
I think the repeated instances of checking for this variable is indicating we need to do some refactoring here. I already had a hunch that we had done too much sharding and fragmentation in this architecture, and this repeated logic seems to confirm that theory. In v1 we had the extreme that had a particularly monolithic class hierarchy that was difficult to read because we had everything in one place, but in this code we appear to have went to the opposite extreme by trying to make every trait do exactly one thing, but that might have been too granular and results in problems like this.

I wonder if you could come up with a design that makes it such that we only have to switch once. One example of this could be to have a single method in this provider class that creates an init container "bundle", which can be a case class / struct of all the things that's required to set up one of these init containers. Then, we can have a switch that provides an `Option` of one of these bundles, and either returns all of the components, or none of them.

We don't actually have to merge the classes themselves, though in some cases that might make sense. We particularly need to isolate the `SparkPodInitContainerBootstrap` because it's a component that's re-used for both the driver pod and the executor pods. But there might be other cases where it makes more sense to combine some of the classes.

I'm fairly open to any ideas on how to make this code better, but can you see what can be done here?